event_routing_backends/backends/events_router.py

```diff
@@ -28,7 +28,95 @@ def __init__(self, processors=None, backend_name=None):
+        """
+        Create host_configurations for the given host and router.
+        """
+        host['host_configurations'] = {}
+        host['host_configurations'].update({'url': router.route_url})
+        host['host_configurations'].update({'auth_scheme': router.auth_scheme})
+
+        if router.auth_scheme == RouterConfiguration.AUTH_BASIC:
+            host['host_configurations'].update({'username': router.username})
+            host['host_configurations'].update({'password': router.password})
+        elif router.auth_scheme == RouterConfiguration.AUTH_BEARER:
+            host['host_configurations'].update({'auth_key': router.auth_key})
+
+        if router.backend_name == RouterConfiguration.CALIPER_BACKEND:
+            host.update({'router_type': 'AUTH_HEADERS'})
+            if 'headers' in host:
+                host['host_configurations'].update({'headers': host['headers']})
+        elif router.backend_name == RouterConfiguration.XAPI_BACKEND:
+            host.update({'router_type': 'XAPI_LRS'})
+        else:
+            host.update({'router_type': 'INVALID_TYPE'})
+
+        return host
+
+    def prepare_to_send(self, events):
+        """
+        Prepare a list of events to be sent and create a processed, filtered batch for each router.
+        """
+        routers = RouterConfiguration.get_enabled_routers(self.backend_name)
+        business_critical_events = get_business_critical_events()
+        route_events = {}
+
+        # We continue even without routers here to allow logging of statements to happen.
+        # If operators do not wish to log and have no enabled routers they should set XAPI_EVENTS_ENABLED
+        # or CALIPER_EVENTS_ENABLED to false.
+        if not routers:
+            logger.debug('Could not find any enabled router configuration for backend %s', self.backend_name)
+            routers = []
+
+        for event in events:
+            try:
+                event_name = event['name']
+            except TypeError as exc:
+                raise ValueError('Expected event as dict but {type} was given.'.format(type=type(event))) from exc
+
+            try:
+                logger.debug(
+                    'Processing edx event "{}" for router with backend {}'.format(event_name, self.backend_name)
+                )
+
+                processed_event = self.process_event(event)
+            except (EventEmissionExit, ValueError):
+                logger.error(
+                    'Could not process edx event "%s" for backend %s\'s router',
+                    event_name,
+                    self.backend_name,
+                    exc_info=True
+                )
+                continue
+
+            logger.debug(
+                'Successfully processed edx event "%s" for router with backend %s. Processed event: %s',
+                event_name,
+                self.backend_name,
+                processed_event
+            )
+
+            for router in routers:
```

Comment:
Here we batch events into a dictionary of lists, where the key is the primary key of the RouterConfiguration. Later we use these lists as the batches to send to each downstream LRS.